overlord/hookstate/context.go

```diff
@@ -127,3 +129,46 @@ func (c *Context) Get(key string, value interface{}) error {
 
 	return nil
 }
+
+// Cached returns the cached value associated with the provided key.
+// It returns nil if there is no entry for key.
+func (c *Context) Cached(key interface{}) interface{} {
+	contextCache := c.task.State().Cached("hook-context")
+	cache, ok := contextCache.(map[interface{}]interface{})
+	if !ok {
+		return nil
+	}
+
+	return cache[key]
+}
+
+// Cache associates value with key. The cached value is not persisted.
+func (c *Context) Cache(key, value interface{}) {
+	contextCache := c.task.State().Cached("hook-context")
+	cache, ok := contextCache.(map[interface{}]interface{})
+	if !ok {
+		cache = make(map[interface{}]interface{})
+	}
+	cache[key] = value
+	c.task.State().Cache("hook-context", cache)
+}
+
+// OnDone requests the provided finalizer to be run once the context knows it's
+// complete.
+func (c *Context) OnDone(finalizer func() error) {
+	c.finalizer = finalizer
+}
+
+// Done is called to notify the context that its hook has exited successfully.
+func (c *Context) Done() error {
+	if c.finalizer != nil {
+		return c.finalizer()
+	}
+	return nil
+}
+
+// NewTransaction returns a getter/setter for snap configs that runs all
+// operations on a copy of the system configuration until committed.
+func (c *Context) NewTransaction() *transaction.Transaction {
```

Comment:
The problem is that this needs to be based on the state, of which the Context has sole control (we made it that way so handlers can't write directly to the state). Who should be responsible for making these, then? Originally it was the configmanager, which was perfect, but that doesn't work so well here.